kernel: fix message queue put front issue and increase message queue test coverage for put front

TaiJuWu opened the pull request:
A new test covers the behavior of putting a message into the message queue
when it is full.
The reason is that when the buffer is full, Thread A gets pended (blocked).
If Thread B later calls the get function, it will unpend Thread A,
allowing it to resume and put the message into the queue.
In this situation, we need to know whether Thread A should
continue with put to front or put to end.

related: https://github.com/zephyrproject-rtos/zephyr/pull/92865

TaiJuWu committed:
test: add config to test msgq  put front featrue

Add a config `CONFIG_TEST_MSGQ_PUT_FRONT` in mesgq_test to
enable message queue put front feature.

Also, add `test_msgq_pending` to test message queue behavior when
the caller is going to put and put front.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

TaiJuWu commented on include/zephyr/kernel/thread.h:
```diff
@@ -74,6 +74,10 @@ struct _thread_base {
 	/* thread state */
 	uint8_t thread_state;
 
```
If possible, there is any space to put this flag?
Ideally, I don’t want to increase thread memory space for a single bit.

cfriedt replied:
I bet @peter-mitsis would probably know a good place to store this information.

cfriedt commented:
@TaiJuWu - no harm in squashing the two test commits into one.

peter-mitsis reviewed (commented):
A few minor comments, but there is something that I wish to draw attention to ensure that it does not slip beneath notice.

As I understand it, the underlying concern behind this PR is that the 'insert to the front of the queue' property was not being preserved when the previously blocked thread unblocks. As it stands with this proposed solution, it is mostly preserved. I shall attempt to describe what I mean. (This was the reason why had requested the "same behavior" comment in the k_msgq_put_front() description.)

A thread on a pend queue is subject to priority sorting. Thus, if the message queue has enough room for N items AND if at least N higher priority threads block on the message queue then at least one of them will get processed before the thread that wanted to be inserted at the front gets processed. This is probably not that big of an issue, and it can be argued that we already have a sort of precedence where the FIFO nature of the message queue is already partially violated with the same priority sorting in the wait queue.

As I write this, it occurs to me that there is another method available to preserve the 'front' property of a message to at least the same extent as we preserve the FIFO property of a normal message--we introduce a front wait queue.

struct k_msgq {
    wait_q_t wait_q;
    ...
    wait_q_t front_q;
};

When front-putting a message, if the buffer is full, then we insert into the `front_q`. When getting a message, check the `front_q` first. If the `front_q` was empty, then we carry on as we would have normally done. I can see this having some performance impact, so if we go for this method, we might want a Kconfig option to enable/disable it.

andyross reviewed (commented):
Not loving messing with the thread struct for the benefit (only) of msgq, at all.  The thread struct is a mess already, and msgq is one of our more obscure IPC mechanisms.

The inserting thread obviously knows which side of the queue to insert the struct on.  Why can't it figure it out on its own?

TaiJuWu commented:
Hi @andyross @peter-mitsis ,

Thanks for your comments, 
there are two options we can choose, one is from @peter-mitsis and another is from my new idea.

1. see https://github.com/zephyrproject-rtos/zephyr/pull/93751#pullrequestreview-3063632220
2. we can see `put_front` as a urgent behavior so we don't allow it make thread go to sleep.

Both options would not to change thread controller block.
Which one do you like?

TaiJuWu commented:
> Not loving messing with the thread struct for the benefit (only) of msgq, at all. The thread struct is a mess already, and msgq is one of our more obscure IPC mechanisms.
> 
> The inserting thread obviously knows which side of the queue to insert the struct on. Why can't it figure it out on its own?
> 
> If a thread was pended, it will lose this information so we need to record it.

The problem is we miss the information which put function is called.

andyross commented:
Alternative suggestion: make put_front() a non-blocking API.  Eliminate the timeout and treat it as a K_NO_WAIT with an error return indicating "full".   App code would be more complicated, but IMHO manageable.

TaiJuWu committed:
kernel: message does not execute correct put front behavior

When the buffer is full, Thread A gets pended (blocked).
If Thread B later calls the get function, it will unpend Thread A,
allowing it to resume and put the message into the queue.
In this situation, we need to know whether Thread A should
continue with put to front or put to end.

In order to resolve this issue, we don't allow set timeout
parameter for `k_msgq_put_front` and this parameter is always
`K_NO_WAIT`.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

TaiJuWu commented:
@andyross @peter-mitsis @alexpaschoaletto thank for your review, I have updated PR, please take a look.

peter-mitsis reviewed (dismissed):
I don't know why the timeslice reset test failed as that is unrelated to this.

andyross commented on include/zephyr/kernel.h:
```diff
@@ -4820,18 +4820,19 @@ __syscall int k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t time
  * by this function.
  *
+ * @note We consider put front an urgent action, so we assume the timeout parameter
+ * is always K_NO_WAIT.
```
There, uh, is no timeout parameter. :)

I'd phrase this as "k_msgq_put_front() does not block".

andyross commented on include/zephyr/kernel.h:
```diff
@@ -4820,18 +4820,19 @@ __syscall int k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t time
  *
  * @retval 0 Message sent.
  * @retval -ENOMSG Returned without waiting or queue purged.
  * @retval -EAGAIN Waiting period timed out.
```
Doesn't return -EAGAIN anymore

andyross reviewed (commented):
Nitpicks.  Feel free to assume a +1 from me once addressed

JarmouniA reviewed (dismissed):
No need to add me as co-author in the 2nd commit.

TaiJuWu commented:
Only update commit message from feedback https://github.com/zephyrproject-rtos/zephyr/pull/93751#pullrequestreview-3103323603
